runtime/Java/src/org/antlr/v4/runtime/atn/ATNSerializer.java

```diff
@@ -65,141 +75,116 @@ public ATNSerializer(ATN atn, List<String> tokenNames) {
 	 *
 	 *  Convenient to pack into unsigned shorts to make as Java string.
 	 */
-	public IntegerList serialize(String language) {
-		IntegerList data = new IntegerList();
+	public IntegerList serialize() {
+		addPreamble();
+		int nedges = addEdges();
+		addNonGreedyStates();
+		addPrecedenceStates();
+		addRuleStatesAndLexerTokenTypes();
+		addModeStartStates();
+		Map<IntervalSet, Integer> setIndices = null;
+		if ( language.equals("Java") ) {
+			setIndices = addSetsEncoded16bits();
+		}
+		else {
+			setIndices = addSets();
+		}
+		addEdges(nedges, setIndices);
+		addDecisionStartStates();
+		addLexerActions();
+
+		if ( language.equals("Java") ) {
+			for (int i = 1; i < data.size(); i++) {
+				int value = data.get(i);
+				if (value < Character.MIN_VALUE || value > Character.MAX_VALUE) {
+					throw new UnsupportedOperationException("Serialized ATN data element " +
+							value + " element " + i + " out of range " + (int) Character.MIN_VALUE + ".." + (int) Character.MAX_VALUE);
+				}
+
+				// Shift by 2, to avoid inefficient modified utf-8 and coding done by java class files
+				data.set(i, (value + 2) & 0xFFFF);
+			}
+		}
```

Comment:
Can we move this transform to `SerializedATN`? Remove range check and split 32-bit integers on two 16-bit during writing. Especially considering now we have `SerializedATN` and `SerializedJavaATN`.

Reply:
We must think alike because I tried that yesterday and gave up when I couldn't immediately get the deserialization tests working. haha.  I'm going to try again today because it is a good idea to separate this.  I will also try a quick combined 16+32 encoding for java following your ideas.